### PR TITLE
add pos_weight for SigmoidBinaryCrossEntropyLoss

### PR DESCRIPTION
**Pull request by eureka7mt**
## Description ##
Add pos_weight for SigmoidBinaryCrossEntropyLoss.
A value `pos_weights > 1` decreases the false negative count, hence increasing the recall.
Conversely setting `pos_weights < 1` decreases the false positive count and increases the precision.
This can be seen from the fact that `pos_weight` is introduced as a multiplicative coefficient for the positive targets term in the loss expression:
    label * -log(sigmoid(pred)) * pos_weight + (1 - label) * -log(1 - sigmoid(pred))

It's adopted from [tensorflow's implementation](https://github.com/tensorflow/tensorflow/blob/r1.12/tensorflow/python/ops/nn_impl.py#L187)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by roywei** [comment]
@eureka7mt Thanks for the contribution, could you add a unit test for this case?

**Comment by roywei** [comment]
@mxnet-label-bot add[Gluon, pr-awaiting-review]

File: python/mxnet/gluon/loss.py [hunk 1]
```diff
@@ -194,6 +194,14 @@ class SigmoidBinaryCrossEntropyLoss(Loss):
         L = - \sum_i {label}_i * \log({pred}_i) +
             (1 - {label}_i) * \log(1 - {pred}_i)
 
+    A value `pos_weights > 1` decreases the false negative count, hence increasing
```
Comment (anirudhacharya) [inline python/mxnet/gluon/loss.py]: please update the formulas above with the new `pos_weight` parameter. The documentation should specify how the formula changes when `from_sigmoid` is True and False and where in the formula `pos_weight` is used.

File: python/mxnet/gluon/loss.py [hunk 2]
```diff
@@ -227,13 +238,20 @@ def __init__(self, from_sigmoid=False, weight=None, batch_axis=0, **kwargs):
         super(SigmoidBinaryCrossEntropyLoss, self).__init__(weight, batch_axis, **kwargs)
         self._from_sigmoid = from_sigmoid
 
-    def hybrid_forward(self, F, pred, label, sample_weight=None):
+    def hybrid_forward(self, F, pred, label, sample_weight=None, pos_weight=None):
         label = _reshape_like(F, label, pred)
         if not self._from_sigmoid:
             # We use the stable formula: max(x, 0) - x * z + log(1 + exp(-abs(x)))
-            loss = F.relu(pred) - pred * label + F.Activation(-F.abs(pred), act_type='softrelu')
+            if pos_weight is None:
```
Comment (anirudhacharya) [inline python/mxnet/gluon/loss.py]: can default `pos_weight` be set to 1 since it is a multiplicative factor and update the formulas accordingly. Then we can do away with this `if else` statement.

Comment (eureka7mt) [reply python/mxnet/gluon/loss.py]: Well.Set pos_weight=1 is OK.I just want to write it In this format like 'sample_weight=None'.

File: tests/python/unittest/test_loss.py [hunk 3]
```diff
@@ -420,6 +420,37 @@ def test_poisson_nllloss_mod():
             initializer=mx.init.Normal(sigma=0.1), eval_metric=mx.metric.Loss(),
             optimizer='adam')
     assert mod.score(data_iter, eval_metric=mx.metric.Loss())[0][1] < 0.05
+    
+@with_seed()
+def test_bce_loss_with_pos_weight():
+    #Suppose it's a multi-label classification
+    N = 20
```
Comment (anirudhacharya) [inline tests/python/unittest/test_loss.py]: please fetch a random number here for N

**Comment by stu1130** [comment]
@eureka7mt Could you fix the Trailing whitespace issue?

**Comment by vandanavk** [comment]
@eureka7mt could you re-trigger the CI?

**Comment by ankkhedia** [comment]
@eureka7mt Could you please look into the CI failures?

**Comment by anirudhacharya** [comment]
@mxnet-label-bot update [pr-awaiting-merge]

File: python/mxnet/gluon/loss.py [hunk 4]
```diff
@@ -227,13 +234,14 @@ def __init__(self, from_sigmoid=False, weight=None, batch_axis=0, **kwargs):
...
         self._from_sigmoid = from_sigmoid
 
-    def hybrid_forward(self, F, pred, label, sample_weight=None):
+    def hybrid_forward(self, F, pred, label, pos_weight=1, sample_weight=None):
         label = _reshape_like(F, label, pred)
         if not self._from_sigmoid:
             # We use the stable formula: max(x, 0) - x * z + log(1 + exp(-abs(x)))
-            loss = F.relu(pred) - pred * label + F.Activation(-F.abs(pred), act_type='softrelu')
+            log_weight = 1 + (pos_weight - 1) * label
+            loss = pred - pred*label + log_weight*(F.Activation(-F.abs(pred), act_type='softrelu') + F.relu(-pred))
```
Comment (wkcn) [inline python/mxnet/gluon/loss.py]: Thanks for your contribution!
I am checking this formula.

Comment (wkcn) [reply python/mxnet/gluon/loss.py]: Right.

File: python/mxnet/gluon/loss.py [hunk 5]
```diff
@@ -227,13 +234,14 @@ def __init__(self, from_sigmoid=False, weight=None, batch_axis=0, **kwargs):
         super(SigmoidBinaryCrossEntropyLoss, self).__init__(weight, batch_axis, **kwargs)
         self._from_sigmoid = from_sigmoid
 
-    def hybrid_forward(self, F, pred, label, sample_weight=None):
+    def hybrid_forward(self, F, pred, label, pos_weight=1, sample_weight=None):
         label = _reshape_like(F, label, pred)
         if not self._from_sigmoid:
             # We use the stable formula: max(x, 0) - x * z + log(1 + exp(-abs(x)))
```
Comment (wkcn) [inline python/mxnet/gluon/loss.py]: pos_weight will drop the performance when it is equal to 1.
Could you please add the following if-else statement?
```python
            if pos_weight == 1:
                # We use the stable formula: max(x, 0) - x * z + log(1 + exp(-abs(x)))
                loss = F.relu(pred) - pred * label + F.Activation(-F.abs(pred), act_type='softrelu')
            else:
                # We use the stable formula: x - x * z + (1 + z * pos_weight - z) * (log(1 + exp(-abs(x))) + max(-x, 0))
                log_weight = 1 + (pos_weight - 1) * label
                loss = pred - pred*label + log_weight*(F.Activation(-F.abs(pred), act_type='softrelu') + F.relu(-pred))
```

Thanks!

Comment (wkcn) [reply python/mxnet/gluon/loss.py]: Thanks for the contribution!
By the way, don't forget to add your name to CONTRIBUTORS.md

File: CONTRIBUTORS.md [hunk 6]
```diff
@@ -20,52 +20,6 @@ Contributors of Apache MXNet (incubating)
 MXNet has been developed by a community of people who are interested in large-scale machine learning and deep learning.
 Everyone is more than welcomed to is a great way to make the project better and more accessible to more users.
 
-Committers
```
Comment (szha) [inline CONTRIBUTORS.md]: List is outdated and doesn't seem necessary to be maintained.

Comment (eureka7mt) [reply CONTRIBUTORS.md]: I see,I made a mistake when changing it using notepad++.I will fix it

Comment (eureka7mt) [reply CONTRIBUTORS.md]: @wkcn I just worked for another thing and didn't find that the if-else statement wasn't added in the new version.Thanks for your correction.

**Comment by eureka7mt** [comment]
Add the if-else statement make an error.Though the default value of pos_weight is set to be 1,the pos_weight is an (1,N) NDArray.

**Comment by wkcn** [comment]
@eureka7mt 
I see. Since pos_weight is a tensor, it is better to  default pos_weight=None as you wrote before. I will update it.

**Comment by eureka7mt** [comment]
@wkcn It could be a scalar for classifying a single class.But for multi-class and multi-label classifying,it should be a tensor.Because in this situation,the number of positive examples and negative examples isn't same for each class.
And in pytorch,it's also defined as a tensor.See the [pytorch docs](https://pytorch.org/docs/master/nn.html?highlight=binary_cross_entropy_with_logits#torch.nn.functional.binary_cross_entropy_with_logits).

**Comment by wkcn** [comment]
I change the order of SigmoidBinaryCrossEntropyLoss inputs from `(self, F, pred, label, pos_weight=None, sample_weight=None)` to `(self, F, pred, label, sample_weight=None, pos_weight=None)`, since we need the compatibilty for other projects

**Comment by wkcn** [comment]
Sorry that I trigger the sanity problem.
Could someone please help me solve it? Thanks!

**Comment by eureka7mt** [comment]
Maybe the broadcast_mul isn't necessary.I think that a NDArray * a NDArray will do broadcast_mul automatically.

**Comment by wkcn** [comment]
@eureka7mt I think we may pass Symbol into SigmoidBinaryCrossEntropyLoss.

**Comment by wkcn** [comment]
The PR has been merged.
Thanks for your contribution: )
